File: solutions/typescript/crypto-square/1/crypto-square.ts
export class Crypto {
  private normalized: string;
  private encoded: string;

  constructor(plainText: unknown) {
    // Normalize: downcase and remove non-alphanumeric characters
    const text = String(plainText).toLowerCase().replace(/[^a-z0-9]/g, "");
    this.normalized = text;
    this.encoded = this.encode(text);
  }

  get ciphertext(): string {
    return this.encoded;
  }

  private encode(text: string): string {
    const length = text.length;
    if (length === 0) return '';

    // Determine rectangle size
    const sqrt = Math.sqrt(length);
    let rows = Math.floor(sqrt);
    let cols = Math.ceil(sqrt);
    if (rows * cols < length) {
      rows = cols;
    }

    // Break into rows and pad last row if necessary
    const grid: string[] = [];
    for (let i = 0; i < length; i += cols) {
      let row = text.slice(i, i + cols);
      if (row.length < cols) {
        row = row.padEnd(cols, ' ');
      }
      grid.push(row);
    }

    // Read down columns
    const chunks: string[] = [];
    for (let c = 0; c < cols; c++) {
      let chunk = '';
      for (let r = 0; r < rows; r++) {
        const row = grid[r] || ''.padEnd(cols, ' ');
        chunk += row[c] || ' ';
      }
      chunks.push(chunk);
    }

    // Join with spaces
    return chunks.join(' ');
  }
}

File: solutions/python/crypto-square/1/crypto_square.py
import math

def cipher_text(plain_text):
    # Step 1: Normalize the text
    normalized = ''.join(c.lower() for c in plain_text if c.isalnum())
    
    if not normalized:
        return ""

    # Step 2: Determine rectangle size
    length = len(normalized)
    r = int(math.floor(math.sqrt(length)))
    c = int(math.ceil(math.sqrt(length)))
    
    if r * c < length:
        r += 1

    # Step 3: Create the grid
    grid = [normalized[i:i+c] for i in range(0, length, c)]

    # Step 4: Read columns to encode
    encoded_chunks = []
    for col in range(c):
        chunk = ''.join(row[col] if col < len(row) else ' ' for row in grid)
        encoded_chunks.append(chunk)
    
    # Step 5: Return result with space-separated chunks
    return ' '.join(encoded_chunks)
